addons/website_sale/static/src/js/website_sale_recently_viewed.js

```diff
@@ -0,0 +1,185 @@
+odoo.define('website_sale.recently_viewed', function (require) {
+
+var concurrency = require('web.concurrency');
+var config = require('web.config');
+var core = require('web.core');
+var publicWidget = require('web.public.widget');
+var utils = require('web.utils');
+var wSaleUtils = require('website_sale.utils');
+
+var qweb = core.qweb;
+
+publicWidget.registry.productsRecentlyViewedSnippet = publicWidget.Widget.extend({
+    selector: '.s_wsale_products_recently_viewed',
+    xmlDependencies: ['/website_sale/static/src/xml/website_sale_recently_viewed.xml'],
+    events: {
+        'click .js_add_cart': '_onAddToCart',
+    },
+
+    /**
+     * @constructor
+     */
+    init: function () {
+        this._super.apply(this, arguments);
+        this._dp = new concurrency.DropPrevious();
+        this.uniqueId = _.uniqueId('o_carousel_recently_viewed_products_');
+        this._onResizeChange = _.debounce(this._addCarousel, 100);
+    },
+    /**
+     * @override
+     */
+    start: function () {
+        this._dp.add(this._fetch()).then(this._render.bind(this));
+        $(window).resize(() => {
+            this._onResizeChange();
+        });
+        return this._super.apply(this, arguments);
+    },
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * @private
+     */
+    _fetch: function () {
+        return this._rpc({
+            route: '/shop/products/recently_viewed',
+        });
+    },
+    /**
+     * @private
+     */
+    _render: function (res) {
+        if (!res) {
+            return;
+        }
+        var products = res['products'];
+        if (products && products.length) {
+            var mobileProducts = [], webProducts = [], productsTemp = [];
+            _.each(products, function (product) {
+                if (productsTemp.length === 4) {
+                    webProducts.push(productsTemp);
+                    productsTemp = [];
+                }
+                productsTemp.push(product);
+                mobileProducts.push([product]);
+            });
+            productsTemp.length && webProducts.push(productsTemp);
```

Comment:
maybe use an if ? to be more explicit in this case (and in python is standard, not sure in js even if it is the same ;))